libs/wire-subsystems/src/Wire/InvitationCodeStore.hs

```diff
@@ -0,0 +1,124 @@
+-- This file is part of the Wire Server implementation.
+--
+-- Copyright (C) 2022 Wire Swiss GmbH <opensource@wire.com>
+--
+-- This program is free software: you can redistribute it and/or modify it under
+-- the terms of the GNU Affero General Public License as published by the Free
+-- Software Foundation, either version 3 of the License, or (at your option) any
+-- later version.
+--
+-- This program is distributed in the hope that it will be useful, but WITHOUT
+-- ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+-- FOR A PARTICULAR PURPOSE. See the GNU Affero General Public License for more
+-- details.
+--
+-- You should have received a copy of the GNU Affero General Public License along
+-- with this program. If not, see <https://www.gnu.org/licenses/>.
+{-# LANGUAGE RecordWildCards #-}
+{-# LANGUAGE StrictData #-}
+{-# LANGUAGE TemplateHaskell #-}
+
+module Wire.InvitationCodeStore where
+
+import Control.Monad.Trans.Maybe (MaybeT (MaybeT, runMaybeT))
+import Data.Id (InvitationId, TeamId, UserId)
+import Data.Json.Util (UTCTimeMillis)
+import Data.Range (Range)
+import Database.CQL.Protocol (Record (..), TupleType, recordInstance)
+import Imports
+import Polysemy
+import Polysemy.TinyLog (TinyLog)
+import System.Logger.Message qualified as Log
+import URI.ByteString
+import Wire.API.Team.Invitation (Invitation (inviteeEmail))
+import Wire.API.Team.Invitation qualified as Public
+import Wire.API.Team.Role (Role, defaultRole)
+import Wire.API.User (EmailAddress, InvitationCode, Name)
+import Wire.Arbitrary (Arbitrary, GenericUniform (..))
+import Wire.Sem.Logger qualified as Log
+import Wire.Timeout
+
+data StoredInvitation = MkStoredInvitation
+  { teamId :: TeamId,
```

Comment:
TODO: Private constructor?

Reply:
Please let’s not do this. Just name the constructor something that makes it clear like “UnsafeMk…”, if we must?

Reply:
That might also be an idea 🤔